File: js/src/earley.js
// Recognition (engine §4) and the terms and conditions it evaluates
// (engine §10).

import { GencmuError } from "./errors.js";
import { tagKey, tagUnion, tagIntersection, sameTagNames, strongTag, weakTag, tagSet, compareCodePoints } from "./tags.js";

// Interns tag sets, so that an item names a captured part's tags by number.
export class TagInterner {
  constructor() {
    this.sets = [];
    this.ids = new Map();
  }
  intern(tags) {
    const key = tagKey(tags);
    let id = this.ids.get(key);
    if (id === undefined) {
      id = this.sets.length;
      this.sets.push(tags);
      this.ids.set(key, id);
    }
    return id;
  }
  get(id) {
    return this.sets[id];
  }
}

// What a parse and every nested parse it starts share.
export class ParseContext {
  constructor(lowered, tokens, sourceText, unicode) {
    this.lowered = lowered;
    this.tokens = tokens;
    this.sourceText = sourceText;
    this.unicode = unicode;
    this.interner = new TagInterner();
    this.nested = new Map();
    this.inProgress = new Set();
  }
}

class Item {
  constructor(production, dot, origin, slots, key) {
    this.production = production;
    this.dot = dot;
    this.origin = origin;
    this.slots = slots;
    this.key = key;
    this.tagId = -1;
    this.edges = [];
    this.end = -1;
  }
  get complete() {
    return this.dot === this.production.rhs.length;
  }
}

class ChartSet {
  constructor(position) {
    this.position = position;
    this.items = [];
    this.index = new Map();
    this.queue = [];
    this.head = 0;
    this.waiting = new Map();
    this.nullable = new Map();
  }
}

// Runs the recognizer over tokens[start, end) with `rule` as the start rule.
export function recognize(context, rule, start, end) {
  const { lowered, tokens } = context;
  const sets = [];
  for (let position = start; position <= end; position++) sets.push(new ChartSet(position));
  const setAt = (position) => sets[position - start];

  const add = (set, production, dot, origin, slots, edge, tagId) => {
    const key = slotKey(production.id, dot, origin, slots);
    let item = set.index.get(key);
    if (item) {
      if (!item.edges.some((existing) => sameEdge(existing, edge))) item.edges.push(edge);
      return;
    }
    item = new Item(production, dot, origin, slots, key);
    item.end = set.position;
    item.tagId = tagId;
    item.edges.push(edge);
    set.items.push(item);
    set.index.set(key, item);
    set.queue.push(item);
    const next = production.rhs[dot];
    if (next && !next.terminal) {
      if (!set.waiting.has(next.name)) set.waiting.set(next.name, []);
      set.waiting.get(next.name).push(item);
    }
    if (dot === production.rhs.length && origin === set.position) {
      if (!set.nullable.has(production.lhs)) set.nullable.set(production.lhs, []);
      set.nullable.get(production.lhs).push(item);
    }
  };

  const predict = (set, name) => {
    for (const production of lowered.byLhs.get(name) || []) {
      const slots = emptySlots(production);
      if (!conditionsHold(context, production, -1, slots)) continue;
      const tagId = production.rhs.length === 0 ? completeTags(context, production, slots) : -1;
      add(set, production, 0, set.position, slots, SEED, tagId);
    }
  };

  // The item advanced over its next symbol, which spans [from, to) and was
  // built by `child`, or read as a token; null when a condition fails.
  const advance = (item, from, to, child) => {
    const production = item.production;
    let slots = item.slots;
    const captureIndex = production.captures.findIndex((capture) => capture.index === item.dot);
    if (captureIndex >= 0) {
      slots = slots.slice();
      const tags = child ? child.tagId : context.interner.intern(tokens[from].tags);
      slots[captureIndex] = [from, to, tags];
    }
    if (!conditionsHold(context, production, item.dot, slots)) return null;
    const dot = item.dot + 1;
    const tagId = dot === production.rhs.length ? completeTags(context, production, slots) : -1;
    return { dot, slots, tagId };
  };

  predict(setAt(start), rule);
  for (let position = start; position <= end; position++) {
    const set = setAt(position);
    while (set.head < set.queue.length) {
      const item = set.queue[set.head++];
      const next = item.production.rhs[item.dot];
      if (!next) {
        const origin = setAt(item.origin);
        for (const waiting of origin.waiting.get(item.production.lhs) || []) {
          const advanced = advance(waiting, item.origin, position, item);
          if (advanced) {
            add(set, waiting.production, advanced.dot, waiting.origin, advanced.slots,
              { kind: "complete", previous: waiting, child: item }, advanced.tagId);
          }
        }
      } else if (!next.terminal) {
        predict(set, next.name);
        for (const done of set.nullable.get(next.name) || []) {
          const advanced = advance(item, position, position, done);
          if (advanced) {
            add(set, item.production, advanced.dot, item.origin, advanced.slots,
              { kind: "complete", previous: item, child: done }, advanced.tagId);
          }
        }
      } else if (position < end && tokens[position].tags.has(next.name)) {
        const advanced = advance(item, position, position + 1, null);
        if (advanced) {
          add(setAt(position + 1), item.production, advanced.dot, item.origin, advanced.slots,
            { kind: "scan", previous: item, token: position, terminal: next.name }, advanced.tagId);
        }
      }
    }
  }
  return { sets, start, end, setAt };
}

const SEED = { kind: "seed" };

function sameEdge(left, right) {
  if (left.kind !== right.kind) return false;
  if (left.kind === "seed") return true;
  if (left.kind === "scan") return left.previous === right.previous && left.token === right.token && left.terminal === right.terminal;
  return left.previous === right.previous && left.child === right.child;
}

function emptySlots(production) {
  return production.captures.map(() => null);
}

function slotKey(id, dot, origin, slots) {
  let key = id + "," + dot + "," + origin;
  for (const slot of slots) key += slot ? "," + slot[0] + ":" + slot[1] + ":" + slot[2] : ",-";
  return key;
}

// The completed items of `rule` spanning [start, end).
export function rootItems(chart, rule) {
  return chart.setAt(chart.end).items.filter((item) =>
    item.complete && item.origin === chart.start && item.production.lhs === rule);
}

function conditionsHold(context, production, readyAt, slots) {
  for (const { condition, readyAt: at } of production.conditions) {
    if (at !== readyAt) continue;
    const scope = new ChartScope(context, production, slots);
    if (!holds(context, condition, scope)) return false;
  }
  return true;
}

function completeTags(context, production, slots) {
  if (!production.tags) return context.interner.intern(tagSet());
  const scope = new ChartScope(context, production, slots);
  return context.interner.intern(asTagSet(evaluate(context, production.tags, scope)));
}

class ChartScope {
  constructor(context, production, slots) {
    this.context = context;
    this.production = production;
    this.slots = slots;
  }
  capture(name) {
    const index = this.production.captures.findIndex((capture) => capture.name === name);
    const slot = this.slots[index];
    return { start: slot[0], end: slot[1], tags: this.context.interner.get(slot[2]) };
  }
}

// A span value: [start, end) of the stage's tokens, and the tags of the
// captured constituent if it is a whole capture.
function spanOf(context, span, scope) {
  if (span.capture !== undefined) return scope.capture(span.capture);
  if (span.call === "head" || span.call === "tail" || span.call === "last") {
    const inner = spanOf(context, span.args[0], scope);
    const { start, end } = inner;
    if (span.call === "head") return { start, end: Math.min(start + 1, end) };
    if (span.call === "tail") return { start: Math.min(start + 1, end), end };
    return { start: Math.max(end - 1, start), end };
  }
  throw new GencmuError("grammar", `expected a span, found ${JSON.stringify(span)}`);
}

export function phonemesOf(tokens, start, end) {
  let result = "";
  for (let index = start; index < end; index++) result += tokens[index].phonemes || "";
  return result;
}

export function textOf(context, start, end) {
  if (start >= end) return "";
  const from = context.tokens[start].source[0];
  const to = context.tokens[end - 1].source[1];
  return context.sourceText.slice(from, to).join("");
}

function tokensTags(tokens, start, end) {
  let result = tagSet();
  for (let index = start; index < end; index++) result = tagUnion(result, tokens[index].tags);
  return result;
}

export function evaluate(context, term, scope) {
  if (term.literal !== undefined) return { string: term.literal };
  if (term.weak !== undefined) return { tags: weakTag(term.weak) };
  if (term.emptySet) return { tags: tagSet() };
  if (term.set) return { tags: term.set.reduce((acc, item) => tagUnion(acc, asTagSet(evaluate(context, item, scope))), tagSet()) };
  if (term.union) return { tags: term.union.reduce((acc, item) => tagUnion(acc, asTagSet(evaluate(context, item, scope))), tagSet()) };
  if (term.intersection) {
    const [first, ...rest] = term.intersection.map((item) => asTagSet(evaluate(context, item, scope)));
    return { tags: rest.reduce((acc, item) => tagIntersection(acc, item), first) };
  }
  if (term.call !== undefined) {
    const args = term.args;
    switch (term.call) {
      case "phonemes": {
        const span = spanOf(context, args[0], scope);
        return { string: phonemesOf(context.tokens, span.start, span.end) };
      }
      case "text": {
        const span = spanOf(context, args[0], scope);
        return { string: textOf(context, span.start, span.end) };
      }
      case "lowercase": {
        const inner = evaluate(context, args[0], scope);
        return { string: context.unicode.lowercase(asString(inner)) };
      }
      case "words": {
        const span = spanOf(context, args[0], scope);
        return { list: phonemesOf(context.tokens, span.start, span.end).split(" ").filter((word) => word !== "") };
      }
      case "tags": {
        const span = spanOf(context, args[0], scope);
        if (args.length === 2) return { tags: nestedTags(context, args[1].rule, span.start, span.end) };
        if (span.tags && args[0].capture !== undefined) return { tags: span.tags };
        return { tags: tokensTags(context.tokens, span.start, span.end) };
      }
      case "classes": {
        const span = spanOf(context, args[0], scope);
        const tags = span.tags && args[0].capture !== undefined ? span.tags : tokensTags(context.tokens, span.start, span.end);
        const result = tagSet();
        for (const [tag, strong] of tags) {
          const first = tag.charCodeAt(0);
          if (first >= 0x41 && first <= 0x5a) result.set(tag, strong);
        }
        return { tags: result };
      }
      default:
        throw new GencmuError("grammar", `unknown function ${term.call}`);
    }
  }
  if (term.capture !== undefined) {
    const span = scope.capture(term.capture);
    return { tags: span.tags };
  }
  throw new GencmuError("grammar", `unknown term ${JSON.stringify(term)}`);
}

function asTagSet(value) {
  if (value.tags) return value.tags;
  if (value.string !== undefined) return strongTag(value.string);
  return tagSet(value.list.map((item) => [item, true]));
}

function asString(value) {
  if (value.string !== undefined) return value.string;
  throw new GencmuError("grammar", "expected a string");
}

export function holds(context, condition, scope) {
  if (condition.any) return condition.any.some((item) => holds(context, item, scope));
  if (condition.not) return !holds(context, condition.not, scope);
  if (condition.matches !== undefined) {
    const span = spanOf(context, condition.matches, scope);
    return nestedMatches(context, condition.rule, span.start, span.end);
  }
  const left = evaluate(context, condition.left, scope);
  const right = evaluate(context, condition.right, scope);
  switch (condition.op) {
    case "=":
    case "≠": {
      let equal;
      if (left.string !== undefined && right.string !== undefined) equal = left.string === right.string;
      else equal = sameTagNames(asTagSet(left), asTagSet(right));
      return equal === (condition.op === "=");
    }
    case "∈":
    case "∉": {
      const needle = asString(left);
      let member;
      if (right.list) member = right.list.includes(needle);
      else if (right.tags) member = right.tags.has(needle);
      else member = right.string === needle;
      return member === (condition.op === "∈");
    }
    case "⊆": {
      const small = asTagSet(left);
      const large = asTagSet(right);
      for (const tag of small.keys()) if (!large.has(tag)) return false;
      return true;
    }
    default:
      throw new GencmuError("grammar", `unknown comparison ${condition.op}`);
  }
}

// The key under which a nested parse's answer is remembered: everything a
// nested parse can observe (engine §4).
function nestedKey(context, kind, rule, start, end) {
  let key = kind + "\u0001" + rule + "\u0001" + textOf(context, start, end);
  for (let index = start; index < end; index++) {
    const token = context.tokens[index];
    key += "\u0001" + tagKey(token.tags) + "\u0002" + token.text + "\u0002" + (token.phonemes || "");
  }
  return key;
}

function nested(context, kind, rule, start, end, compute) {
  const key = nestedKey(context, kind, rule, start, end);
  if (context.nested.has(key)) return context.nested.get(key);
  const circular = nestedKey(context, "parse", rule, start, end);
  if (context.inProgress.has(circular)) {
    throw new GencmuError("grammar",
      `a condition asks whether ${JSON.stringify(textOf(context, start, end))} parses as ${rule} from inside the parse of that span as ${rule}: ` +
      `the grammar defines ${rule} by its own negation over the same text`, { rule });
  }
  context.inProgress.add(circular);
  try {
    const chart = recognize(context, rule, start, end);
    const answer = compute(chart);
    context.nested.set(key, answer);
    return answer;
  } finally {
    context.inProgress.delete(circular);
  }
}

function nestedMatches(context, rule, start, end) {
  return nested(context, "matches", rule, start, end, (chart) => rootItems(chart, rule).length > 0);
}

function nestedTags(context, rule, start, end) {
  return nested(context, "tags", rule, start, end, (chart) =>
    rootItems(chart, rule).reduce((acc, item) => tagUnion(acc, context.interner.get(item.tagId)), tagSet()));
}

// The furthest position the parse reached, and what could have been read
// there, with the rules that could have read it.
export function rejectionOf(chart, lowered) {
  let position = chart.end;
  while (position > chart.start && chart.setAt(position).items.length === 0) position--;
  const expected = new Map();
  for (const item of chart.setAt(position).items) {
    const next = item.production.rhs[item.dot];
    if (!next || !next.terminal) continue;
    if (!expected.has(next.name)) expected.set(next.name, new Set());
    expected.get(next.name).add(item.production.owner);
  }
  void lowered;
  return {
    position,
    expected: [...expected.keys()].sort(compareCodePoints).map((terminal) => ({
      terminal,
      rules: [...expected.get(terminal)].sort(compareCodePoints),
    })),
  };
}

File: lib/go/earley.go
package gencmu

// The recognizer (engine §4): an Earley parser whose items carry, for each
// capture before the dot, the captured part's span and tag set, and which
// evaluates each condition as soon as the item has read the last capture it
// mentions. Every item records how it was reached, which is the packed
// forest the ranking (rank.go) reads.

type capVal struct {
	start, end int32
	tags       int32
}

type itemKey struct {
	prod   *production
	dot    int32
	origin int32
	caps   [4]capVal
}

type item struct {
	itemKey
	set   int32
	links []link
}

// link is one way an item was reached: its predecessor (nil for dot 1 from a
// prediction) and the child it read, a token or a completed constituent.
type link struct {
	prev *item
	sym  *symNode // nil for a read
	tok  int32    // for a read: the token, relative to the recognizer's base
	term int32
}

// symNode is every completed item of one rule over one span with one tag set.
type symNode struct {
	rule       int32
	start, end int32
	tags       *tagset
	items      []*item
}

type symKey struct {
	rule, origin, tags int32
}

type eset struct {
	items     []*item
	index     map[itemKey]*item
	waiting   map[int32][]*item
	predicted map[int32]bool
	syms      map[symKey]*symNode
	empties   map[int32][]*symNode
}

type recognizer struct {
	run      *stageRun
	g        *lowered
	base, n  int
	sets     []*eset
	furthest int
}

func (r *recognizer) set(k int) *eset {
	for len(r.sets) <= k {
		r.sets = append(r.sets, &eset{index: map[itemKey]*item{}})
	}
	return r.sets[k]
}

// recognize runs the recognizer over tokens [base, base+n) of the stage with
// start as the start rule. While it runs, the input that initial() tests
// begins at base, and a nested recognition sets its own.
func (run *stageRun) recognize(g *lowered, start int32, base, n int) *recognizer {
	outer := run.inputStart
	run.inputStart = base
	defer func() { run.inputStart = outer }()
	r := &recognizer{run: run, g: g, base: base, n: n}
	s0 := r.set(0)
	r.predict(s0, 0, start)
	for k := 0; k <= n && k < len(r.sets); k++ {
		s := r.sets[k]
		if len(s.items) > 0 {
			r.furthest = k
		}
		for i := 0; i < len(s.items); i++ {
			r.process(k, s.items[i])
		}
	}
	return r
}

// predict adds the items of a rule's productions at k, except those whose
// first symbol is a terminal the next token lacks, which could never
// advance; expected() accounts for them in a rejection.
func (r *recognizer) predict(s *eset, k int, rule int32) {
	if s.predicted[rule] {
		return
	}
	if s.predicted == nil {
		s.predicted = map[int32]bool{}
	}
	s.predicted[rule] = true
	for _, p := range r.g.rules[rule].prods {
		if len(p.rhs) > 0 && p.rhs[0].term && !r.canRead(k, p.rhs[0].id) {
			continue
		}
		if !r.predictable(p, k) {
			continue
		}
		r.add(k, itemKey{prod: p, origin: int32(k)}, link{}, false)
	}
}

func (r *recognizer) canRead(k int, term int32) bool {
	if k >= r.n {
		return false
	}
	_, ok := r.run.tagsets[r.base+k].has(r.g.terminals[term])
	return ok
}

// predictable checks the conditions of a production that mention no
// capture, which hold or fail at prediction at k, as do those of a
// production with no symbols that mention $, over the empty span there.
func (r *recognizer) predictable(p *production, k int) bool {
	if len(p.predictConds) > 0 {
		var caps [4]capVal
		var tags *tagset
		if len(p.rhs) == 0 {
			tags = r.completedTags(p, &caps, int32(k), int32(k))
		}
		ev := r.run.evaluator(r.g, r.captureFunc(p, &caps, int32(k), int32(k), tags))
		ok := true
		for _, c := range p.predictConds {
			if !ev.cond(c) {
				ok = false
				break
			}
		}
		if !ok {
			return false
		}
	}
	return true
}

func (r *recognizer) add(k int, key itemKey, l link, hasLink bool) {
	s := r.set(k)
	it := s.index[key]
	if it == nil {
		it = &item{itemKey: key, set: int32(k)}
		s.index[key] = it
		s.items = append(s.items, it)
	}
	if hasLink {
		it.links = append(it.links, l)
	}
}

func (r *recognizer) process(k int, it *item) {
	p := it.prod
	s := r.sets[k]
	if int(it.dot) < len(p.rhs) {
		sym := p.rhs[it.dot]
		if sym.term {
			if k < r.n {
				ts := r.run.tagsets[r.base+k]
				if _, ok := ts.has(r.g.terminals[sym.id]); ok {
					r.advance(it, k+1, capVal{int32(k), int32(k + 1), ts.id}, link{prev: it, tok: int32(k), term: sym.id})
				}
			}
			return
		}
		if s.waiting == nil {
			s.waiting = map[int32][]*item{}
		}
		s.waiting[sym.id] = append(s.waiting[sym.id], it)
		r.predict(s, k, sym.id)
		for _, c := range s.empties[sym.id] {
			r.advance(it, k, capVal{c.start, c.end, c.tags.id}, link{prev: it, sym: c})
		}
		return
	}
	// Complete.
	ts := r.completedTags(p, &it.caps, it.origin, int32(k))
	key := symKey{p.lhs, it.origin, ts.id}
	c := s.syms[key]
	if c != nil {
		c.items = append(c.items, it)
		return
	}
	c = &symNode{rule: p.lhs, start: it.origin, end: int32(k), tags: ts, items: []*item{it}}
	if s.syms == nil {
		s.syms = map[symKey]*symNode{}
	}
	s.syms[key] = c
	if int(it.origin) == k {
		if s.empties == nil {
			s.empties = map[int32][]*symNode{}
		}
		s.empties[p.lhs] = append(s.empties[p.lhs], c)
	}
	waiters := r.sets[it.origin].waiting[p.lhs]
	for i := 0; i < len(waiters); i++ {
		w := waiters[i]
		r.advance(w, k, capVal{c.start, c.end, ts.id}, link{prev: w, sym: c})
	}
}

// advance moves an item over its next symbol, read over cv, into set k,
// unless a condition triggered there fails.
func (r *recognizer) advance(it *item, k int, cv capVal, l link) {
	p := it.prod
	key := it.itemKey
	pos := int(key.dot)
	if slot := p.capSlot[pos]; slot >= 0 {
		key.caps[slot] = cv
	}
	key.dot++
	var whole *tagset
	for _, c := range p.conds {
		if c.trigger == int(key.dot) {
			// A condition on $ is evaluated once the item is complete,
			// with the tags its production's tag term gives it (§4).
			if c.whole && whole == nil {
				whole = r.completedTags(p, &key.caps, key.origin, int32(k))
			}
			if !r.run.evaluator(r.g, r.captureFunc(p, &key.caps, key.origin, int32(k), whole)).cond(c.cond) {
				return
			}
		}
	}
	if l.prev != nil && l.prev.dot == 0 {
		l.prev = nil // a predicted item has no derivation of its own
	}
	r.add(k, key, l, true)
}

// captureFunc gives an item's captures; $ spans [origin, end) and has the
// given tags, nil while they are being computed, when a term cannot read
// them (§9).
func (r *recognizer) captureFunc(p *production, caps *[4]capVal, origin, end int32, whole *tagset) func(string) (spanVal, bool) {
	return func(name string) (spanVal, bool) {
		if name == "" {
			return spanVal{a: r.base + int(origin), b: r.base + int(end), whole: whole != nil, tags: whole}, true
		}
		slot, ok := p.slotOf[name]
		if !ok {
			return spanVal{}, false
		}
		cv := caps[slot]
		return spanVal{a: r.base + int(cv.start), b: r.base + int(cv.end), whole: true, tags: r.run.ps.in.all[cv.tags]}, true
	}
}

// completedTags is the constituent tags of a production's item over
// [origin, end) with the given captures (engine §4).
func (r *recognizer) completedTags(p *production, caps *[4]capVal, origin, end int32) *tagset {
	in := r.run.ps.in
	switch {
	case p.tags != nil:
		return r.run.evaluator(r.g, r.captureFunc(p, caps, origin, end, nil)).tagsOf(p.tags)
	case p.implicit:
		return in.all[caps[p.capSlot[0]].tags]
	}
	return in.empty()
}

// accepted lists the completed start-rule constituents over the whole input.
func (r *recognizer) accepted(start int32) []*symNode {
	if len(r.sets) <= r.n {
		return nil
	}
	var out []*symNode
	for key, c := range r.sets[r.n].syms {
		if key.rule == start && key.origin == 0 {
			out = append(out, c)
		}
	}
	sortSyms(out)
	return out
}

func sortSyms(s []*symNode) {
	for i := 1; i < len(s); i++ {
		for j := i; j > 0 && s[j].tags.key < s[j-1].tags.key; j-- {
			s[j], s[j-1] = s[j-1], s[j]
		}
	}
}
